source/custom/nixnet_service.custom.cpp

```diff
@@ -431,46 +442,114 @@ ::grpc::Status NiXnetService::GetProperty(::grpc::ServerContext* context, const
         std::vector<nxDatabaseRef_t> property_value_vector(number_of_elements, 0U);
         nxDatabaseRef_t* property_value = static_cast<nxDatabaseRef_t*>(property_value_vector.data());
         status = library_->GetProperty(session, property_id, property_size, property_value);
-        response->mutable_db_ref_array()->mutable_db_ref()->Clear();
-        response->mutable_db_ref_array()->mutable_db_ref()->Reserve(number_of_elements);
-        std::transform(
-            property_value_vector.begin(),
-            property_value_vector.end(),
-            google::protobuf::RepeatedFieldBackInserter(response->mutable_db_ref_array()->mutable_db_ref()),
-            [&](auto x) {
-              auto init_lambda = [&]() {
-                return std::make_tuple(status, x);
-              };
-              uint32_t session_id{};
-              status = nx_database_ref_t_resource_repository_->add_dependent_session("", init_lambda, initiating_session_id, session_id);
-              nidevice_grpc::Session dependent_session{};
-              dependent_session.set_id(session_id);
-              return dependent_session;
-            });
-        if (!status_ok(status)) {
-          response->set_status(status);
-          return ::grpc::Status::OK;
+        if (status_ok(status)) {
+          response->mutable_db_ref_array()->mutable_db_ref()->Clear();
+          response->mutable_db_ref_array()->mutable_db_ref()->Reserve(number_of_elements);
+          auto cleanup_lambda = [&](nxDatabaseRef_t id) {};
+          std::transform(
+              property_value_vector.begin(),
+              property_value_vector.end(),
+              google::protobuf::RepeatedFieldBackInserter(response->mutable_db_ref_array()->mutable_db_ref()),
+              [&](auto x) {
+                auto init_lambda = [&]() {
+                  return std::make_tuple(status, x);
+                };
+                uint32_t session_id{};
+                status = nx_database_ref_t_resource_repository_->add_session("", init_lambda, cleanup_lambda, session_id);
+                nidevice_grpc::Session session{};
+                session.set_id(session_id);
+                return session;
+              });
+        }
+        break;
+      }
+      case dev_ref_: {
+        auto initiating_session_id = session_repository_->access_session_id(session_grpc_session.id(), session_grpc_session.name());
+        auto init_lambda = [&]() {
+          nxSessionRef_t property_value;
+          status = library_->GetProperty(session, property_id, property_size, &property_value);
+          return std::make_tuple(status, property_value);
+        };
+        uint32_t session_id = 0;
+        // We are adding it to session_repository_ and not to "device repository", because devices don't have a close API,
+        // so it makes sense to tie their lifetime with sessions lifetime.
+        status = session_repository_->add_dependent_session("", init_lambda, initiating_session_id, session_id);
```

Comment:
note: We can use add_dependent_session here and below because both sessions belong to same repository